File: android/src/main/kotlin/com/cashier/pos_printer/PosPrinterPlugin.kt
package com.cashier.pos_printer

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import com.cashier.pos_printer.printing.Printer
import com.cashier.pos_printer.printing.senraise.SenraisePrinterHelper
import com.cashier.pos_printer.printing.sunmi.SunmiPrintHelper
import com.cashier.pos_printer.printing.telpo.TelpoPrintHelper
import com.cashier.pos_printer.utils.DeviceFeaturesUtils
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding

/** PosPrinterPlugin */
class PosPrinterPlugin : FlutterPlugin, PosPrinter, ActivityAware {

    private lateinit var printer: Printer
    private lateinit var context: Context

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        PosPrinter.setUp(flutterPluginBinding.binaryMessenger, this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        PosPrinter.setUp(binding.binaryMessenger, null)
    }


    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        init(binding.activity)
    }

    override fun onDetachedFromActivityForConfigChanges() {
        if (::context.isInitialized)
            printer.deInitPrinter(context)
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        init(binding.activity)
    }

    override fun onDetachedFromActivity() {
        if (::context.isInitialized)
            printer.deInitPrinter(context)
    }

    private fun init(context: Context) {
        printer = when {
            DeviceFeaturesUtils.isSenraise() -> SenraisePrinterHelper()
            DeviceFeaturesUtils.isSunmi() -> SunmiPrintHelper()
            DeviceFeaturesUtils.isTelpo() -> TelpoPrintHelper()
            else -> SunmiPrintHelper()
        }
        printer.init(context)
    }

    override fun start() {
        printer.start()
    }

    override fun printText(text: String, textSize: Double, isBold: Boolean, isItalic: Boolean) {
        printer.printText(
            text = text,
            textSize = textSize.toFloat(),
            isBold = isBold,
            isItalic = isItalic
        )
    }

    override fun printTable(
        texts: List<String>,
        width: List<Long>,
        align: List<Long>,
        fontSize: Long,
    ) {
        printer.printTable(
            texts.toTypedArray(),
            width.map { it.toInt() }.toIntArray(),
            align.map { it.toInt() }.toIntArray(),
            fontSize.toInt(),
        )
    }

    override fun printBitmap(bitmap: ByteArray) {
        val decodedBitmap: Bitmap = BitmapFactory.decodeByteArray(bitmap, 0, bitmap.size)
        printer.printBitmap(decodedBitmap)
    }

    override fun feedPaper(lines: Long) {
        printer.feedPaper(lines.toInt())
    }

    override fun setAlign(align: Alignments) {
        printer.setAlign(align.raw)
    }

    override fun getPrinterSize(): PrinterSize {
        return when (printer.getPrinterSize()) {
            com.cashier.pos_printer.printing.numnum.PrinterSize.SIZE_58 -> PrinterSize.SIZE_58MM
            com.cashier.pos_printer.printing.numnum.PrinterSize.SIZE_80 -> PrinterSize.SIZE_80MM
        }
    }

    override fun sendTextToLcdDigital(text: String) {
        printer.sendTextToLcdDigital(text)
    }

    override fun sendImageLcdDigital(bitmap: ByteArray) {
        val decodedBitmap: Bitmap = BitmapFactory.decodeByteArray(bitmap, 0, bitmap.size)
        printer.sendImageToLcdDigital(decodedBitmap)
    }

    override fun openDrawer() {
        printer.openDrawer()
    }

    override fun cutPaper() {
        printer.cutPaper()
    }

    override fun escPosCommandExe(commands: ByteArray) {
        printer.escPosCommandExe(commands)
    }

    override fun getPrinterStatus(): PrinterStatus {
        return printer.getStatus()
    }

    override fun deInitPrinter() {
        if (::context.isInitialized)
            printer.deInitPrinter(context)
    }

    override fun getDeviceManufacture(): DeviceManufacture {
        return when {
            DeviceFeaturesUtils.isSenraise() -> DeviceManufacture.SENRAISE
            DeviceFeaturesUtils.isSunmi() -> DeviceManufacture.SUNMI
            DeviceFeaturesUtils.isTelpo() -> DeviceManufacture.TELPO
            else -> DeviceManufacture.UNKNOWN
        }
    }

    override fun release() {
        printer.release()
    }

}

File: pigeons/pos_printer_plugin.dart
import 'package:pigeon/pigeon.dart';

@ConfigurePigeon(PigeonOptions(
  dartOut: 'lib/src/pos_printer.g.dart',
  kotlinOptions: KotlinOptions(
    package: 'com.cashier.pos_printer',
  ),
  kotlinOut: 'android/src/main/kotlin/com/cashier/pos_printer/PosPrinter.kt',
  dartPackageName: 'pos_printer',
))
enum Alignments { left, center, right }

enum PrinterSize { size_58mm, size_80mm }

enum PrinterStatus { normal, outOfPaper, overHeating, coverOpen, generalError }

enum DeviceManufacture { sunmi, senraise, telpo, unknown }

@HostApi()
abstract class PosPrinter {
  /// start printing
  /// used in case of printers only working with commit mode
  /// like telpo printers
  void start();

  /// print text
  /// [text] string to be printed
  /// [isBold] to set text to bold
  /// [isItalic] to set text to italic
  /// [textSize] 27f is preferred
  void printText(String text, double textSize, bool isBold, bool isItalic);

  /// Print a row of a table
  /// [texts] text table to be printed
  /// [width] determine the width of text
  /// [align] determine the alignment of text in the row
  /// [fontSize] fontSize of the printed text
  void printTable(
      List<String> texts, List<int> width, List<int> align, int fontSize);

  /// Print an image
  void printBitmap(Uint8List bitmap);

  /// Due to the distance between the paper hatch and the print head,
  /// the paper needs to be fed out manually
  void feedPaper(int lines);

  /// Set printer alignment
  void setAlign(Alignments align);

  /// Get printer head size 58mm or 80mm
  PrinterSize getPrinterSize();

  /// Send String to display on 7 segments display for sunmi d3 mini
  void sendTextToLcdDigital(String text);

  /// Send Image to display on LCD
  void sendImageLcdDigital(Uint8List bitmap);

  /// open cash drawer for supported device
  void openDrawer();

  ///cut the paper after printing
  void cutPaper();

  ///send esc/pos [commands] as bytes
  void escPosCommandExe(Uint8List commands);

  ///return the current status of the printer
  PrinterStatus getPrinterStatus();

  ///release printer after quitting app
  void deInitPrinter();

  ///return pos manufacture sunmi,senraise,telpo and
  ///unknown in case of unsupported brand
  DeviceManufacture getDeviceManufacture();

  /// release printer after printing complete
  /// works with telpo and pos that support commit mode
  void release();
}
